### ISSUE TITLE
shallow: props are not updated in case shouldComponentUpdate returns false

### ISSUE THREAD

**Issue by atsikov** [airbnb/enzyme#1599]
### Current behavior

Using `shouldComponentUpdate` I want to prevent update on prop change and wait until state is changed due to some asynchronous action.
When I use `setState` on `shallow` mounted component to test expected result, it uses props value from last component update, not the last `setProps` call.

This test works properly using `mount`, or if I remove `shouldComponentUpdate`.

### Expected behavior

`render` uses previously changed props, even if actual update for these props was prevented using `shouldComponentUpdate`.

### Your environment

jest@22.4.2

#### API

shallow

#### Version

| library         | version
| ---------------- | -------
| Enzyme           |  3.3.0
| React            |  16.2.0

#### Adapter

enzyme-adapter-react-16@1.1.1
react-test-renderer@16.2.0

### React behaviour

https://jsfiddle.net/xkyxhnL8/1/

#### ShallowTest.tsx
```
import * as React from "react";

export interface ShallowTestProps {
    foo: string;
    bar?: string;
}

export interface ShallowTestState {
    asyncBar?: string;
}

export class ShallowTest extends React.Component<ShallowTestProps, ShallowTestState> {
    constructor(props: ShallowTestProps) {
        super(props);
        this.state = {};
    }

    public componentWillReceiveProps(nextProps: ShallowTestProps) {
        if (this.props.bar !== nextProps.bar) {
            // some async call to modify state.asyncBar as a result
        }
    }

    public shouldComponentUpdate(nextProps: ShallowTestProps, nextState: ShallowTestState): boolean {
        return this.state.asyncBar !== nextState.asyncBar ||
            this.props.foo !== nextProps.foo;
    }

    public render() {
        return (<div>{this.props.bar}; {this.state.asyncBar}</div>);
    }
}
```

#### ShallowTest.spec.tsx
```
import * as React from "react";
import { shallow, mount } from "enzyme";
import { ShallowTest } from "./ShallowTest";

it("shallow: should render proper async prop value", () => {
    const wrapper = shallow(<ShallowTest foo="init" bar="first" />);

    // this should update component because of foo prop change
    wrapper.setProps({ foo: "update", bar: "second" });

    // this should not update component but should update props
    wrapper.setProps({ bar: "third" });

    // this should allow update component
    wrapper.setState({ asyncBar: "third-async" });

    expect(wrapper.html()).toBe("<div>third; third-async</div>");
});

it("mount: should render proper async prop value", () => {
    const wrapper = mount(<ShallowTest foo="init" bar="first" />);

    // this should update component because of foo prop change
    wrapper.setProps({ foo: "update", bar: "second" });

    // this should not update component but should update props
    wrapper.setProps({ bar: "third" });

    // this should allow update component
    wrapper.setState({ asyncBar: "third-async" });

    expect(wrapper.html()).toBe("<div>third; third-async</div>");
});
```

#### Jest output
```
 FAIL  src/ShallowTest.spec.tsx
  ✕ shallow: should render proper async prop value (19ms)
  ✓ mount: should render proper async prop value (11ms)

  ● shallow: should render proper async prop value

    expect(received).toBe(expected) // Object.is equality

    Expected value to be:
      "<div>third; third-async</div>"
    Received:
      "<div>second; third-async</div>"

      15 |     wrapper.setState({ asyncBar: "third-async" });
      16 |
    > 17 |     expect(wrapper.html()).toBe("<div>third; third-async</div>");
      18 | });
      19 |
      20 | it("mount: should render proper async prop value", () => {

      at Object.<anonymous> (src/ShallowTest.spec.tsx:17:28)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
```

**Comment by ljharb** [comment]
A PR with a failing test case (and in actual JS) would be very helpful.

**Comment by ljharb** [comment]
Closing, but happy to reopen with a test case.

**Comment by atsikov** [comment]
I checked it once again. Looks like the issue was fixed in enzyme@3.4.0.
